### PR TITLE
Reimplement suite_teardown mechanism

### PR DESCRIPTION
**Pull request by RaymondHuang210129**
The current implementation of `suite_teardown` callback iscalled after `etest` runs through all test cases. However, if the `failfast` is enabled and a test case does not pass, the test aborts without calling the `suite_teardown` callback.

This PR reimplements `suite_teardown` mechanism so that it will be triggered at the end of the test, for all following scenario:
1. All tests passes
2. Some tests fails without `failfast` enabled
3. Test aborted when a test fails with failfast `enabled`
4. Test aborted by receiving an interrupt.

### CODE REVIEW

File: share/etest/runners.sh [hunk 1]
```diff
@@ -108,6 +108,15 @@ run_single_test()
                 suite_setup
             fi
 
+            # Register __suite_teardown function as a trap callback and trigger it when the test receives an interrupt.
+            # If running the last test case, instead of triggering it by interrupt, make this function always triggered
+            # at the end of the test. 
+            if [[ ${testidx} -eq ${testidx_total} ]]; then
+                trap_add __suite_teardown EXIT
+            else
+                trap_add __suite_teardown ${DIE_SIGNALS[@]}
```
Comment (RaymondHuang210129) [inline share/etest/runners.sh:117]: The signal is trigger by either user's interrupt or parent process that sends the signal to this process when `failfast` is set.

Comment (marshall-mcmullen) [reply share/etest/runners.sh:117]: This is clever. Using `trap_add` to ensure that the suite teardown function is called is a really elegant solution to this.

File: share/etest/runners.sh [hunk 2]
```diff
@@ -108,6 +108,15 @@ run_single_test()
                 suite_setup
             fi
 
+            # Register __suite_teardown function as a trap callback and trigger it when the test receives an interrupt.
+            # If running the last test case, instead of triggering it by interrupt, make this function always triggered
+            # at the end of the test.
```
Comment (marshall-mcmullen) [inline share/etest/runners.sh:113]: Nit: Reword this to

|  ... instead of triggering it by interrupt, always trigger it at the end of the test.

File: share/etest/runners.sh [hunk 3]
```diff
@@ -129,16 +138,18 @@ run_single_test()
...
-            if is_function suite_teardown && [[ ${testidx} -eq ${testidx_total} ]]; then
-                etestmsg "Running suite_teardown $(lval testidx testidx_total)"
-                suite_teardown
-            fi
         }
         catch
         {
             rc=$?
+            # If failfast flag is enabled and is not running the last test case, call __suite_teardown in the catch block.
+            # Otherwise, nothing to do at here and __suite_teardown will be triggered at the end of the test.
```
Comment (marshall-mcmullen) [inline share/etest/runners.sh:146]: Nit: Reword this to say

> # Otherwise, nothing to do here as __suite_teardown will be triggered at the end of the test.

File: share/etest/runners.sh [hunk 4]
```diff
@@ -205,6 +216,15 @@ run_single_test()
     fi
 }
 
+# A wrapper function that calls suite_teardown if it is defined by user.
+__suite_teardown()
```
Comment (marshall-mcmullen) [inline share/etest/runners.sh:220]: Ah ha! I was wondering what this was when I saw you calling it up above. This is a nice solution @RaymondHuang210129 !

**Review by marshall-mcmullen — CHANGES_REQUESTED** [review]
Small changes requested in wording of comments but otherwise this looks really elegant.

File: share/etest/runners.sh (hunk 2 above)
Comment (RaymondHuang210129) [reply share/etest/runners.sh]: fixed, thanks for the suggestion!

File: share/etest/runners.sh (hunk 3 above)
Comment (RaymondHuang210129) [reply share/etest/runners.sh]: fixed as well.

**Review by marshall-mcmullen — APPROVED** [review]
This looks great now @RaymondHuang210129. I'm going to go ahead and approve and merge this and let the CI tests run. If all goes well we can upgrade what we use to this new version.
